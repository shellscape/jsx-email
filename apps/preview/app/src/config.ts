interface AppConfig {
  excludeGlob: string | undefined;
  relativePath: string;
}

export const config: AppConfig = {
  excludeGlob: import.meta.env.VITE_JSXE_EXCLUDE_GLOB,
  // FIXME: with v2, remove the `define` replacement. use envars for better back compat and
  // assymetric updates
  // @ts-expect-error
  relativePath: import.meta.env.VITE_JSXE_RELATIVE_PATH || `${__JSX_EMAIL_RELATIVE_PATH__}`
};
