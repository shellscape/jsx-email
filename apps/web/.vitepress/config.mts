import { defineConfig } from 'vitepress';

import { sidebar } from './sidebar.mjs';

// TODO: make a vitepress plugins for pnpm, yarn, npm, etc based on
// https://github.com/vuejs/vitepress/blob/6edc588e5c1f01f50c1e158c705e04c1745db1e0/src/node/markdown/plugins/containers.ts#L56

// https://vitepress.dev/reference/site-config
export default defineConfig({
  appearance: 'force-dark',
  base: '/',
  cleanUrls: true,
  description: 'Build emails with a delightful DX',
  head: [
    [
      'link',
      {
        rel: 'shortcut icon',
        href: `data:image/svg+xml;charset=UTF-8,%3csvg width='126' height='113' viewBox='0 0 126 113' fill='none' xmlns='http://www.w3.org/2000/svg'%3e%3cpath fill-rule='evenodd' clip-rule='evenodd' d='M0.199951 50V109V113H4.19995H121.8H125.8V109V50H117.8V105H8.19995V50H0.199951Z' fill='%2364595C'/%3e%3cpath d='M0 53.429V47.4258L48.3069 22.8124V32.4176L11.2516 50.2773L11.5517 49.677V51.1778L11.2516 50.5775L48.3069 68.4372V78.0424L0 53.429Z' fill='%2364595C'/%3e%3cpath d='M79.4367 0L54.6832 92H46.582L71.3356 0H79.4367Z' fill='%2364595C'/%3e%3cpath d='M126 53.429L77.6931 78.0424V68.4372L114.748 50.5775L114.448 51.1778V49.677L114.748 50.2773L77.6931 32.4176V22.8124L126 47.4258V53.429Z' fill='%2364595C'/%3e%3c/svg%3e `
      }
    ]
  ],
  ignoreDeadLinks: true,
  markdown: {
    theme: {
      dark: 'slack-dark',
      light: 'slack-ochin'
    }
  },
  outDir: './dist',
  srcDir: 'src',
  themeConfig: {
    docFooter: {
      prev: false,
      next: false
    },
    logo: '/logo.svg',
    // https://vitepress.dev/reference/default-theme-config
    nav: [
      { text: 'Documentation', link: '/docs/introduction' },
      { text: 'Quick Start', link: '/docs/quick-start' },
      { text: 'Contributions', link: '/contributions' },
      { text: 'Email Samples', link: 'http://samples.jsx.email' }
    ],
    search: {
      provider: 'local'
      // FIXME: We're waiting on DocSearch approval
      // provider: 'algolia',
      // options: {
      //   appId: '...',
      //   apiKey: '...',
      //   indexName: '...'
      // }
    },
    sidebar,
    siteTitle: '',
    socialLinks: [
      { icon: 'discord', link: 'https://discord.gg/FywZN57mTg' },
      { icon: 'github', link: 'https://github.com/shellscape/jsx-email' }
    ]
  },
  titleTemplate: 'JSX email â€¢ :title',
  transformPageData(pageData) {
    pageData.frontmatter.head ??= [];
    pageData.frontmatter.head.push(
      ['meta', { name: 'og:description', content: pageData.description }],
      ['meta', { name: 'og:image', content: 'https://jsx.email/og.png' }],
      ['meta', { name: 'og:site_name', content: 'JSX email' }],
      ['meta', { name: 'og:title', content: pageData.title }],
      ['meta', { name: 'og:type', content: 'website' }],
      ['meta', { name: 'twitter:card', content: 'summary_large_image' }],
      ['meta', { name: 'twitter:description', content: pageData.description }],
      ['meta', { name: 'twitter:image', content: 'https://jsx.email/og.png' }],
      ['meta', { name: 'twitter:title', content: pageData.title }]
    );
  }
});
