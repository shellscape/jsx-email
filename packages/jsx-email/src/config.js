import { AssertionError } from 'assert';
import { isAbsolute, resolve as resolvePath } from 'node:path';
import { pathToFileURL } from 'node:url';
import chalk from 'chalk';
import { lilconfig } from 'lilconfig';
import { getPluginLog, log } from './log.js';
// @ts-ignore
import { pluginSymbol } from './plugins.js';
const configSymbol = Symbol.for('jsx-email/config');
const toImportSpecifier = (id) => {
    if (id.startsWith('file://'))
        return id;
    if (isAbsolute(id))
        return pathToFileURL(id).href;
    if (id.startsWith('.'))
        return pathToFileURL(resolvePath(id)).href;
    return id;
};
export const defaults = {
    logLevel: 'info',
    plugins: [],
    render: {
        disableDefaultStyle: false,
        inlineCss: false,
        minify: false,
        plainText: false,
        pretty: false
    }
};
export const globalConfigSymbol = Symbol.for('jsx-email/global/config');
export const current = () => {
    let curr = globalThis[globalConfigSymbol];
    // eslint-disable-next-line no-multi-assign
    if (!curr)
        curr = globalThis[globalConfigSymbol] = { ...defaults };
    return curr;
};
const plugins = {
    inline: '@jsx-email/plugin-inline',
    minify: '@jsx-email/plugin-minify',
    pretty: '@jsx-email/plugin-pretty'
};
const checkSymbol = (plugin, source) => {
    if (!plugin)
        return;
    const oops = `did not have the required 'symbol' property set correctly, and it can't be verified as a valid plugin`;
    const message = source
        ? chalk `{red jsx-email}: A plugin imported from '${source}' ${oops}`
        : chalk `{red jsx-email}: The plugin named '${plugin.name}' ${oops}`;
    if (plugin.symbol !== pluginSymbol) {
        log.error(message);
        throw new AssertionError({ message });
    }
};
const checkName = (plugin, source) => {
    if (!plugin)
        return;
    const oops = `did not have the required 'name' property`;
    const message = source
        ? chalk `{red jsx-email}: A plugin imported from '${source}' ${oops}`
        : chalk `{red jsx-email}: A plugin added to the config ${oops}`;
    if (!plugin.name) {
        log.error(message);
        process.exit(1);
    }
};
const handleImportError = (error, name) => {
    if (error.code === 'ERR_MODULE_NOT_FOUND') {
        log.error(chalk `{red jsx-email}: Tried to import plugin '${name}' but it wasn't found`);
    }
    else {
        log.error(error);
    }
};
// Note: We have to be verbose here so that bundlers pick up on the imports.
// Most of them can't handle dynamically importing these plugins from variables containing
// their names without additional config, and we don't want that burden on users
const importInlinePlugin = async () => {
    try {
        // Note: tshy up to bullshit again with compile errors where there are none
        // @ts-ignore
        const { plugin } = (await import('@jsx-email/plugin-inline'));
        return plugin;
    }
    catch (error) {
        handleImportError(error, '@jsx-email/plugin-inline');
    }
    return null;
};
const importMinifyPlugin = async () => {
    try {
        // @ts-ignore
        const { plugin } = (await import('@jsx-email/plugin-minify'));
        return plugin;
    }
    catch (error) {
        handleImportError(error, '@jsx-email/plugin-minify');
    }
    return null;
};
const importPrettyPlugin = async () => {
    try {
        // @ts-ignore
        const { plugin } = (await import('@jsx-email/plugin-pretty'));
        return plugin;
    }
    catch (error) {
        handleImportError(error, '@jsx-email/plugin-pretty');
    }
    return null;
};
export const defineConfig = async (config = {}) => {
    if (typeof config === 'function') {
        const intermediate = await config();
        return defineConfig(intermediate);
    }
    const mods = structuredClone(defaults);
    if (!config.render?.plainText) {
        // Note: The order of plugins here actually matters for how the doc gets
        // transformed. Changing this ordering may produce undesirable html
        if (config.render?.inlineCss) {
            const inline = await importInlinePlugin();
            if (inline)
                mods.plugins.push(inline);
        }
        if (config.render?.minify) {
            const minify = await importMinifyPlugin();
            if (minify)
                mods.plugins.push(minify);
        }
        if (config.render?.pretty) {
            const pretty = await importPrettyPlugin();
            if (pretty)
                mods.plugins.push(pretty);
        }
        if (config.render?.minify && config.render.pretty) {
            log.warn(chalk `{yellow jsx-email}: Both minify and pretty options are true. Please choose only one.`);
        }
        if ((config.render?.minify || config.render?.pretty) && config.render?.plainText) {
            log.warn(chalk `{yellow jsx-email}: plaintText has been enabled, minify and pretty will have no effect.`);
            mods.plugins = mods.plugins.filter(({ name }) => name !== plugins.minify && name !== plugins.pretty);
        }
    }
    const result = await mergeConfig(mods, config);
    const pluginMap = new Map();
    // Note: make sure we don't have duplicate plugins
    for (const plugin of result.plugins || [])
        pluginMap.set(plugin.name, plugin);
    result.plugins = Array.from(pluginMap, ([, value]) => value);
    for (const plugin of result.plugins || []) {
        checkName(plugin);
        checkSymbol(plugin);
        plugin.log = getPluginLog(plugin.name);
    }
    result.symbol = configSymbol;
    return result;
};
const moduleImport = async (id) => {
    const specifier = toImportSpecifier(id);
    try {
        const mod = await import(specifier);
        return mod;
    }
    catch (e) {
        try {
            if (require)
                return require(id);
            throw RangeError('require is not defined. this is likely due to the ESM build attempting to use require');
        }
        catch (error) {
            if (error.code === 'ERR_REQUIRE_ESM' ||
                (error instanceof SyntaxError &&
                    error.toString().includes('Cannot use import statement outside a module'))) {
                throw e;
            }
            throw error;
        }
    }
};
export const loadConfig = async (startDir) => {
    if (globalThis[globalConfigSymbol])
        return globalThis[globalConfigSymbol];
    const name = 'jsx-email';
    const searchResult = await lilconfig(name, {
        loaders: {
            '.cjs': moduleImport,
            '.js': moduleImport,
            '.mjs': moduleImport
        },
        searchPlaces: [
            `.config/${name}rc.js`,
            `.config/${name}rc.cjs`,
            `.config/${name}rc.mjs`,
            `.config/${name}.config.js`,
            `.config/${name}.config.cjs`,
            `.config/${name}.config.mjs`,
            `.${name}rc.js`,
            `.${name}rc.cjs`,
            `.${name}rc.mjs`,
            `${name}.config.js`,
            `${name}.config.cjs`,
            `${name}.config.mjs`
        ]
    }).search(startDir);
    log.debug('loadConfig â†’', { cwd: process.cwd(), searchResult, startDir });
    const configExports = searchResult?.config ?? {};
    const intermediate = configExports.config instanceof Promise ? await configExports.config : configExports.config;
    const config = intermediate ?? { ...defaults };
    if (config.symbol === configSymbol) {
        globalThis[globalConfigSymbol] = config;
        return config;
    }
    const definedConfig = await defineConfig(config);
    process.env.DOT_LOG_LEVEL = definedConfig.logLevel || 'info';
    setConfig(definedConfig);
    return definedConfig;
};
export const setConfig = (config) => (globalThis[globalConfigSymbol] = config);
// Note: This helps avoid issues with `merge ` because it uses structuredClone and doesn't play nice
// with complex objects'
// Keeping this async in case we need it for the future
export const mergeConfig = async (a, b) => {
    const aPlugins = a.plugins || [];
    const bPlugins = b.plugins || [];
    /* eslint-disable no-param-reassign */
    delete a.plugins;
    delete b.plugins;
    const result = {
        ...a,
        ...b,
        plugins: [...aPlugins, ...bPlugins],
        render: { ...a.render, ...b.render }
    };
    return result;
};
//# sourceMappingURL=config.js.map