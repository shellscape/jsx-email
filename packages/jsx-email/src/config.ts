import { AssertionError } from 'assert';

import chalk from 'chalk';
import { lilconfig } from 'lilconfig';
import type { MethodFactoryLevels } from '@dot/log';

import { getPluginLog, log } from './log.js';
// @ts-ignore
import { pluginSymbol, type JsxEmailPlugin, type PluginInternal } from './plugins.js';
import type { ESBuildOptions, RenderOptions } from './types.js';

export interface JsxEmailConfig {
  esbuild?: ESBuildOptions;
  logLevel?: MethodFactoryLevels;
  plugins: JsxEmailPlugin[];
  render: RenderOptions;
}
export type JsxEmailConfigFn = () => Promise<Partial<JsxEmailConfig>>;
export type DefineConfigOptions = Partial<JsxEmailConfig> | JsxEmailConfigFn;

interface ConfigExports {
  config: Promise<JsxEmailConfig> | Partial<JsxEmailConfig>;
}

const configSymbol = Symbol.for('jsx-email/config');
export const defaults: JsxEmailConfig = {
  logLevel: 'info',
  plugins: [],
  render: {
    disableDefaultStyle: false,
    inlineCss: false,
    minify: false,
    plainText: false,
    pretty: false
  }
};
export const globalConfigSymbol = Symbol.for('jsx-email/global/config');
export const current = (): JsxEmailConfig => {
  let curr = (globalThis as any)[globalConfigSymbol];

  // eslint-disable-next-line no-multi-assign
  if (!curr) curr = (globalThis as any)[globalConfigSymbol] = { ...defaults };

  return curr;
};

const plugins = {
  inline: '@jsx-email/plugin-inline',
  minify: '@jsx-email/plugin-minify',
  pretty: '@jsx-email/plugin-pretty'
};

const checkSymbol = (plugin: JsxEmailPlugin, source?: string) => {
  if (!plugin) return;

  const oops = `did not have the required 'symbol' property set correctly, and it can't be verified as a valid plugin`;
  const message = source
    ? chalk`{red jsx-email}: A plugin imported from '${source}' ${oops}`
    : chalk`{red jsx-email}: The plugin named '${plugin.name}' ${oops}`;

  if (plugin.symbol !== pluginSymbol) {
    log.error(message);
    throw new AssertionError({ message });
  }
};

const checkName = (plugin: JsxEmailPlugin, source?: string) => {
  if (!plugin) return;

  const oops = `did not have the required 'name' property`;
  const message = source
    ? chalk`{red jsx-email}: A plugin imported from '${source}' ${oops}`
    : chalk`{red jsx-email}: A plugin added to the config ${oops}`;

  if (!plugin.name) {
    log.error(message);
    process.exit(1);
  }
};

interface PluginImport {
  plugin: JsxEmailPlugin;
}

const handleImportError = (error: any, name: string) => {
  if (error.code === 'ERR_MODULE_NOT_FOUND') {
    log.error(chalk`{red jsx-email}: Tried to import plugin '${name}' but it wasn't found`);
  } else {
    log.error(error);
  }
};

// Note: We have to be verbose here so that bundlers pick up on the imports.
// Most of them can't handle dynamically importing these plugins from variables containing
// their names without additional config, and we don't want that burden on users
const importInlinePlugin = async () => {
  try {
    // Note: tshy up to bullshit again with compile errors where there are none
    // @ts-ignore
    const { plugin } = (await import('@jsx-email/plugin-inline')) as unknown as PluginImport;
    return plugin;
  } catch (error) {
    handleImportError(error, '@jsx-email/plugin-inline');
  }

  return null;
};

const importMinifyPlugin = async () => {
  try {
    // @ts-ignore
    const { plugin } = (await import('@jsx-email/plugin-minify')) as unknown as PluginImport;
    return plugin;
  } catch (error) {
    handleImportError(error, '@jsx-email/plugin-minify');
  }

  return null;
};

const importPrettyPlugin = async () => {
  try {
    // @ts-ignore
    const { plugin } = (await import('@jsx-email/plugin-pretty')) as unknown as PluginImport;
    return plugin;
  } catch (error) {
    handleImportError(error, '@jsx-email/plugin-pretty');
  }

  return null;
};

export const defineConfig = async (config: DefineConfigOptions = {}): Promise<JsxEmailConfig> => {
  if (typeof config === 'function') {
    const intermediate = await config();
    return defineConfig(intermediate);
  }

  const mods = structuredClone(defaults);

  if (!config.render?.plainText) {
    // Note: The order of plugins here actually matters for how the doc gets
    // transformed. Changing this ordering may produce undesirable html
    if (config.render?.inlineCss) {
      const inline = await importInlinePlugin();
      if (inline) mods.plugins.push(inline);
    }

    if (config.render?.minify) {
      const minify = await importMinifyPlugin();
      if (minify) mods.plugins.push(minify);
    }

    if (config.render?.pretty) {
      const pretty = await importPrettyPlugin();
      if (pretty) mods.plugins.push(pretty);
    }

    if (config.render?.minify && config.render.pretty) {
      log.warn(
        chalk`{yellow jsx-email}: Both minify and pretty options are true. Please choose only one.`
      );
    }
    if ((config.render?.minify || config.render?.pretty) && config.render?.plainText) {
      log.warn(
        chalk`{yellow jsx-email}: plaintText has been enabled, minify and pretty will have no effect.`
      );

      mods.plugins = mods.plugins.filter(
        ({ name }) => name !== plugins.minify && name !== plugins.pretty
      );
    }
  }

  const result = await mergeConfig(mods, config);
  const pluginMap = new Map<string, JsxEmailPlugin>();

  // Note: make sure we don't have duplicate plugins
  for (const plugin of result.plugins || []) pluginMap.set(plugin.name, plugin);

  result.plugins = Array.from(pluginMap, ([, value]) => value);

  for (const plugin of result.plugins || []) {
    checkName(plugin);
    checkSymbol(plugin);
    (plugin as PluginInternal).log = getPluginLog(plugin.name);
  }

  (result as any).symbol = configSymbol;

  return result;
};

const moduleImport = async (id: string) => {
  try {
    const mod = await import(id);
    return mod;
  } catch (e) {
    try {
      return require(id);
    } catch (error) {
      if (
        (error as any).code === 'ERR_REQUIRE_ESM' ||
        (error instanceof SyntaxError &&
          error.toString().includes('Cannot use import statement outside a module'))
      ) {
        throw e;
      }
      throw error;
    }
  }
};

export const loadConfig = async (startDir?: string): Promise<JsxEmailConfig> => {
  if ((globalThis as any)[globalConfigSymbol]) return (globalThis as any)[globalConfigSymbol];

  const name = 'jsx-email';
  const searchResult = await lilconfig(name, {
    loaders: {
      '.cjs': moduleImport,
      '.js': moduleImport,
      '.mjs': moduleImport
    },
    searchPlaces: [
      `.config/${name}rc.js`,
      `.config/${name}rc.cjs`,
      `.config/${name}rc.mjs`,
      `.config/${name}.config.js`,
      `.config/${name}.config.cjs`,
      `.config/${name}.config.mjs`,
      `.${name}rc.js`,
      `.${name}rc.cjs`,
      `.${name}rc.mjs`,
      `${name}.config.js`,
      `${name}.config.cjs`,
      `${name}.config.mjs`
    ]
  }).search(startDir);

  log.debug('loadConfig â†’', { cwd: process.cwd(), searchResult, startDir });

  const configExports: ConfigExports = searchResult?.config ?? {};
  const intermediate =
    configExports.config instanceof Promise ? await configExports.config : configExports.config;
  const config = intermediate ?? { ...defaults };

  if ((config as any).symbol === configSymbol) {
    (globalThis as any)[globalConfigSymbol] = config;
    return config as JsxEmailConfig;
  }

  const definedConfig = await defineConfig(config);

  process.env.DOT_LOG_LEVEL = definedConfig.logLevel || 'info';

  setConfig(definedConfig);

  return definedConfig;
};

export const setConfig = (config: JsxEmailConfig) =>
  ((globalThis as any)[globalConfigSymbol] = config);

// Note: This helps avoid issues with `merge ` because it uses structuredClone and doesn't play nice
// with complex objects'
// Keeping this async in case we need it for the future
export const mergeConfig = async (a: Partial<JsxEmailConfig>, b: Partial<JsxEmailConfig>) => {
  const aPlugins = a.plugins || [];
  const bPlugins = b.plugins || [];

  /* eslint-disable no-param-reassign */
  delete (a as any).plugins;
  delete (b as any).plugins;

  const result = {
    ...a,
    ...b,
    plugins: [...aPlugins, ...bPlugins],
    render: { ...a.render, ...b.render }
  } as JsxEmailConfig;

  return result;
};
