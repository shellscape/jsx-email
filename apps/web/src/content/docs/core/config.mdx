---
title: Configuration
description: Configure JSX-email for your project
type: core
---

<h1>{frontmatter.title}</h1>
<p>{frontmatter.description}</p>

## Configuration File

JSX-email looks for a configuration file when running the `email` CLI command or using the [`render`](/core/render) function. The search starts from the current directory and moves up until it finds either a git repository root directory or the OS's home directory.

### Supported File Names

Configuration files can use any of these names and locations:

```console
.config/jsx-emailrc.js
.config/jsx-emailrc.cjs
.config/jsx-emailrc.mjs
.config/jsx-email.config.js
.config/jsx-email.config.cjs
.config/jsx-email.config.mjs
.jsx-emailrc.js
.jsx-emailrc.cjs
.jsx-emailrc.mjs
jsx-email.config.js
jsx-email.config.cjs
jsx-email.config.mjs
```

:::info
jsx-email does not currently support TypeScript, YAML, or JSON files.
:::

## Usage

### Basic Configuration

Create a configuration file with the required `config` export:

```ts
import { defineConfig } from 'jsx-email/config';

export const config = defineConfig({
  logLevel: 'info',
  plugins: [],
  render: { minify: true }
});
```

The named export `config` is required regardless of how the file is constructed. Note the use of the `defineConfig` import and the `config` named export - while any plain object will be accepted and parsed, the use of `defineConfig` is encouraged for ensuring your config is compliant.

### Intellisense

Enable IDE intellisense using JSDoc type hints:

```ts
/** @type {import('jsx-email/config').JsxEmailConfig} */
export const config = { ... };
```

### Async Configuration

The configuration supports async operations in several ways:

```ts
export const config = (async () => { ... })();
```

Or

```ts
export const config = new Promise((resolve, reject) => { ... })
```

Alternatively, when using the `defineConfig` function, an async function may be passed to it as the first argument:

```ts
export const config = defineConfig(async () => {
  const data = await asyncFunction();
  return {
    // configuration options
  };
});
```

## Configuration Options

### Log Level (optional)

Control the verbosity of console output:

```ts
logLevel?: 'debug' | 'info' | 'warn' | 'error';
```

Default: `info`

### Plugins (optional)

Add plugins to extend functionality:

```ts
plugins?: JsxEmailPlugin[];
```

See [`JsxEmailPlugin` documentation](/core/plugins) for more details.

### Render (optional)

Configure rendering behavior:

```ts
render?: RenderOptions;
```

These options merge with the [`render` function options](/core/render#method-options).
