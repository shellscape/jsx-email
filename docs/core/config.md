---
title: 'Configuration'
description: 'Configuration'
---

## Configuring jsx-email

When running `email` from the command line or programmatically using the [`render`](/docs/core/render) function, jsx-email will attempt to find and use a `jsx-email.config` file. jsx-email will traverse a directory structure looking for config files until it reaches a git repository root directory, or the OS's home directory. This is quite fast, and allows for scoping your configuration file at the filesystem level you'd like to. By default, jsx-email looks for configuration files with the following file extensions and locations:

```
.config/jsx-emailrc.js
.config/jsx-emailrc.cjs
.config/jsx-emailrc.mjs
.config/jsx-email.config.js
.config/jsx-email.config.cjs
.config/jsx-email.config.mjs
.jsx-emailrc.js
.jsx-emailrc.cjs
.jsx-emailrc.mjs
jsx-email.config.js
jsx-email.config.cjs
jsx-email.config.mjs
```

::: info
jsx-email does not currently support TypeScript, YAML, or JSON files.
:::

### Config File

The most basic configuration file might look like this:

```ts
import { defineConfig } from 'jsx-email/config';

export const config = defineConfig({
  render: { minify: true }
});
```

The named export `config` is required regardless of how the file is constructed. Note the use of the `defineConfig` import and the `config` named export - while any plain object will be accepted and parsed, the use of `defineConfig` is encouraged for ensuring your config is compliant.

### Intellisense

Since jsx-email ships with TypeScript typings, you can leverage your IDE's intellisense with jsdoc type hints:

```ts
/** @type {import('jsx-email/config').JsxEmailConfig} */
export const config = { ... };
```

### Async Configuration

The `config` export inherently supports `Promise`. This is helpful when needing to perform other async operations before returning a value to the export. This can be done in a few ways:

```ts
export const config = (async () => { ... })();
```

Or

```ts
export const config = new Promise((resolve, reject) => { ... })
```

Alternatively, when using the `defineConfig` function, an async function may be passed to it as the first argument:

```ts
export const config = defineConfig(async () => {
  const data = await asyncFunction()
  return {
    ...
  }
});
```

### Available Properties

jsx-email configuration files support the following properties:

```ts
logLevel?: 'debug' | 'info' | 'warn' | 'error';
```

_Optional_. Default: `info`. The level at which logs for jsx-email will appear in the terminal / console.

```ts
plugins?: JsxEmailPlugin[];
```

_Optional_. An array of plugins; objects which contain a valid [`JsxEmailPlugin` definition](/docs/core/plugins).

```ts
render?: RenderOptions;
```

_Optional_. The value specified will be merged with the [`render` function options](https://jsx.email/docs/core/render#method-options) when a render takes place. Please see the [`render` documentation](https://jsx.email/docs/core/render#method-options) for more details.
