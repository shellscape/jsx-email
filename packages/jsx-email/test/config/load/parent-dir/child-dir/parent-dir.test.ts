import { loadConfig } from '../../../../../src/config.js';

describe('loadConfig â†’ parent dir', async () => {
  test('loadConfig', async () => {
    const config = await loadConfig(__dirname);
    expect(config).toBeDefined();
    expect(config).toMatchSnapshot();
  });
});
