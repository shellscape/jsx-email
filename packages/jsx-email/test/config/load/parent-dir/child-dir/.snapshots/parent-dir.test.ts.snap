// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`loadConfig → parent dir > loadConfig 1`] = `
{
  "logLevel": "info",
  "plugins": [
    {
      "log": LogLevel {
        "currentLevel": 2,
        "debug": [Function],
        "error": [Function],
        "info": [Function],
        "log": [Function],
        "methodFactory": LogFactory {
          "options": {
            "level": [Function],
            "name": [Function],
            "ready": {
              "promise": Promise {},
              "reject": [Function],
              "resolve": [Function],
            },
            "template": "[{{time}}] jsx-email ∵ root/minify {{level}} ",
            "time": [Function],
          },
          "ready": {
            "promise": Promise {},
            "reject": [Function],
            "resolve": [Function],
          },
          Symbol(log-instance): [Circular],
          Symbol(log-levels): {
            "DEBUG": 1,
            "ERROR": 4,
            "INFO": 2,
            "SILENT": 5,
            "TRACE": 0,
            "WARN": 3,
          },
        },
        "name": "dot-log:∵ root/minify",
        "options": {
          "factory": LogFactory {
            "options": {
              "level": [Function],
              "name": [Function],
              "ready": {
                "promise": Promise {},
                "reject": [Function],
                "resolve": [Function],
              },
              "template": "[{{time}}] jsx-email ∵ root/minify {{level}} ",
              "time": [Function],
            },
            "ready": {
              "promise": Promise {},
              "reject": [Function],
              "resolve": [Function],
            },
            Symbol(log-instance): [Circular],
            Symbol(log-levels): {
              "DEBUG": 1,
              "ERROR": 4,
              "INFO": 2,
              "SILENT": 5,
              "TRACE": 0,
              "WARN": 3,
            },
          },
          "id": "dot-log:∵ root/minify",
          "level": "info",
          "name": "dot-log:∵ root/minify",
          "prefix": undefined,
        },
        "trace": [Function],
        "type": "LogLevel",
        "warn": [Function],
      },
      "name": "root/minify",
      "process": [Function],
      "symbol": Symbol(jsx-email/plugin),
    },
  ],
  "render": {
    "disableDefaultStyle": true,
    "inlineCss": false,
    "minify": true,
    "plainText": false,
    "pretty": false,
  },
  "symbol": Symbol(jsx-email/config),
}
`;
